core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/validation/DiskSpaceValidationService.java

```diff
@@ -0,0 +1,104 @@
+package com.sequenceiq.cloudbreak.service.upgrade.validation;
+
+import java.text.DecimalFormat;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import javax.inject.Inject;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import com.sequenceiq.cloudbreak.common.exception.UpgradeValidationFailedException;
+import com.sequenceiq.cloudbreak.domain.stack.Stack;
+import com.sequenceiq.cloudbreak.orchestrator.host.HostOrchestrator;
+import com.sequenceiq.cloudbreak.orchestrator.model.GatewayConfig;
+import com.sequenceiq.cloudbreak.orchestrator.model.Node;
+import com.sequenceiq.cloudbreak.service.CloudbreakException;
+import com.sequenceiq.cloudbreak.service.GatewayConfigService;
+import com.sequenceiq.cloudbreak.service.resource.ResourceService;
+import com.sequenceiq.cloudbreak.util.StackUtil;
+
+@Component
+public class DiskSpaceValidationService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(DiskSpaceValidationService.class);
+
+    private static final long DIVIDER_TO_MB = 1024L;
+
+    private static final long GB_IN_KB = 1048576;
+
+    private static final double MASTER_NODE_PARCEL_SIZE_MULTIPLIER = 3.5;
+
+    private static final double PARCEL_SIZE_MULTIPLIER = 2.5;
+
+    @Inject
+    private GatewayConfigService gatewayConfigService;
+
+    @Inject
+    private StackUtil stackUtil;
+
+    @Inject
+    private HostOrchestrator hostOrchestrator;
+
+    @Inject
+    private ResourceService resourceService;
+
+    @Inject
+    private ParcelSizeService parcelSizeService;
+
+    public void validateFreeSpaceForUpgrade(Stack stack, String imageCatalogUrl, String imageCatalogName, String imageId) throws CloudbreakException {
+        long parcelSize = parcelSizeService.getAllParcelSize(imageCatalogUrl, imageCatalogName, imageId, stack);
+        Set<Node> nodes = getNodes(stack);
+        Map<String, String> freeDiskSpaceByNodes = getFreeDiskSpaceByNodes(stack, nodes);
+        LOGGER.debug("Required free space for parcels {} KB. Free space by nodes in KB: {}", parcelSize, freeDiskSpaceByNodes);
+        Set<String> notEligibleNodes = getNotEligibleNodes(freeDiskSpaceByNodes, parcelSize, nodes);
+        if (!notEligibleNodes.isEmpty()) {
+            throw new UpgradeValidationFailedException(
```

Comment:
This message is invalid now, it requires more than the actual parcels size by 2.5x and 3.5x more